rfcs/0000-libp2p.md

```diff
@@ -0,0 +1,149 @@
+# libp2p for coda
+
+## Summary
+[summary]: #summary
+
+Coda's networking today uses an [largely-unmaintained Kademlia library written in Haskell](https://github.com/codaprotocol/kademlia). We've been intending to replace it with some other maintained S/Kademlia implementation. We use Kademlia for peer discovery: finding IP address and ports of other protocol participants. On top of this, we build a "gossip net" for broadcasting messages to all participants with a simple flooding protocol. libp2p offers these same primitives plus some useful features like NAT traversal and connection relaying. It also has a Javascript implementation that works in the browser, which will be useful for our SDK.
+
+## Detailed Design
+[detailed-design]: #detailed-design
+
+libp2p itself is a large, flexible, extensible network stack with multiple implementations and multiple options for various functionality. I propose this minimal initial configuration:
+
+
+- On the daemon side, we will use `go-libp2p`. `rust-libp2p` lacks some features (pubsub validation) that we want.
+- For the transport, secio over TCP (optionally with WebSocket). secio is libp2p's custom transport security protocol. I believe the libp2p developers intend to replace secio with TLS 1.3 in the future. Transport security is not yet essential for us, but it's nice to have and will become more important once we start using relays.
```

Comment:
what's a relay?

Reply:
A relay is a server that shuttles messages between two peers. Encryption is important so that the peers can get privacy guarantees for their communications. An example of when it might be needed is when two parties that are behind a symmetric NAT need an intermediate relay to establish communication. In particular, you'd need a TURN server for making this happen.